src/segy/header.py

```diff
@@ -0,0 +1,194 @@
+    """Base class for header transformation strategies."""
+
+    def transform(self, data: NDArray[Any]) -> NDArray[Any]:
+        """Forward transformation implementation."""
+        raise NotImplementedError
+
+    def inverse_transform(self, data: NDArray[Any]) -> NDArray[Any]:
+        """Inverse transformation implementation."""
+        raise NotImplementedError
+
+
+class ScalingStrategy(HeaderTransformationStrategy):
+    """Transform to scale user defined keys in a structured dtype.
+
+    Args:
+        scale_factor: Scalar to use in transformations.
+        keys: List of keys to apply the transformation on.
+    """
+
+    def __init__(self, scale_factor: int | float, keys: list[str]):
+        self.scale_factor = scale_factor
+        self.keys = keys
+
+    def transform(self, data: NDArray[Any]) -> NDArray[Any]:
+        """Forward transformation implementation."""
+        if data.dtype.names is None:
+            msg = f"{self.__class__.__name__} can only work on structured arrays."
+            raise ValueError(msg)
+
+        for key in self.keys:
+            if key in data.dtype.names:
+                data[key] = [vals * self.scale_factor for vals in data[key]]
+        return data
+
+    def inverse_transform(self, data: NDArray[Any]) -> NDArray[Any]:
+        """Inverse transformation implementation."""
+        if data.dtype.names is None:
+            msg = f"{self.__class__.__name__} can only work on structured arrays."
+            raise ValueError(msg)
+
+        for key in self.keys:
+            if key in data.dtype.names:
+                data[key] = [vals / self.scale_factor for vals in data[key]]
+        return data
+
+
+class ByteSwapStrategy(HeaderTransformationStrategy):
+    """Transform to swap bytes in a structured dtype.
+
+    We need to store both source and target, so its reversible.
+
+    Args:
+        source_byteorder: Source byte order.
+        target_byteorder: Desired byte order.
+    """
+
+    def __init__(self, source_byteorder: Endianness, target_byteorder: Endianness):
+        self.source_byteorder = source_byteorder
+        self.target_byteorder = target_byteorder
+
+    def transform(self, data: NDArray[Any]) -> NDArray[Any]:
+        """Swap bytes if target != source. Else it is a no-op."""
+        if self.source_byteorder is not self.target_byteorder:
+            data = data.newbyteorder(self.target_byteorder.symbol).byteswap()
+        return data
+
+    def inverse_transform(self, data: NDArray[Any]) -> NDArray[Any]:
+        """Swap bytes if source != target. Else it is a no-op."""
+        if self.target_byteorder is not self.source_byteorder:
+            data = data.newbyteorder(self.source_byteorder.symbol).byteswap()
+        return data
+
+
+class TransformationPipeline:
+    """Pipeline to chain transforms forward and backward."""
+
+    def __init__(self) -> None:
+        self.transformations: list[HeaderTransformationStrategy] = []
```

Comment:
What if we used something like collections.deque instead of a list to hold the transformations?

Reply:
What would be the advantage here? I don't think it matters that much because we built this list once. Maybe we can benchmark later.